flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/PropertyRead.scala

```diff
@@ -0,0 +1,33 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.table.runtime.aggregate
+
+import org.apache.flink.streaming.api.windowing.windows.Window
+
+/**
+  * Base class for reading a window property. The property will be extracted once and
+  * can be read multiple times.
+  */
+trait PropertyRead[T] extends Serializable {
+
+  def extract(window: Window): Unit
```

Comment:
I think we do not need a generic interface to extract properties. The `extract` method requires a window object but not all properties might be present in the window object. So I am not sure how much generalization is required here.

I think we can implement a `TimeWindowPropertyCollector` that works as follows:
```
class TimeWindowPropertyCollector(start: Int, end: Int) extends Collector[Row] {
  var finalCollector: Collector[Row] = _
  var timeWindow: TimeWindow = _

  override def collect(record: Row): Unit = {

    if (start > -1) {
      record.setField(start, timeWindow.getStart)
    }
    if (end > -1) {
      record.setField(end, timeWindow.getEnd)
    }

    finalCollector.collect(record)
  }

  override def close(): Unit = finalCollector.close()
}
```

If we (ever) need more properties, we can add another wrapping collector or change the design. But at this point I think this is sufficient. 

What do you think?